virtualmachine/gcp/util.go

```diff
@@ -435,3 +442,284 @@ func (svc *googleService) insertSSHKey() error {
+		newRules = append(newRules, &googlecloud.FirewallAllowed{
+			IPProtocol: endpoint.Protocol,
+			Ports:      endpoint.Ports,
+		})
+	}
+
+	return svc.patchFirewall(newRules)
+}
+
+// patchFirewall patches the firewall with the given allowed ports (rules)
+func (svc *googleService) patchFirewall(allowed []*googlecloud.FirewallAllowed) error {
+	newFirewall := &googlecloud.Firewall{
+		Allowed: allowed,
+	}
+	op, err := svc.service.Firewalls.Patch(svc.vm.Project, svc.vm.Firewall,
+		newFirewall).Do()
+	if err != nil {
+		return err
+	}
+	return svc.waitForGlobalOperationReady(op.Name)
+}
+
+// removeFirewallPorts removes ports from a given firewall
+func (svc *googleService) removeFirewallPorts() error {
+	firewall, err := svc.getFirewall()
+	if err != nil {
+		return err
+	}
+	// Initialize new rules with the current one
+	// Refer to this link for structure of firewall resource
+	// https://cloud.google.com/compute/docs/reference/latest/firewalls
+	newRules := firewall.Allowed
+	// Define a map to identify ports to be removed
+	remPorts := make(map[string]map[string]bool)
+	// Define a map to deal with elements where the same protocol
+	// appears again
+	encountered := make(map[string]bool)
+	// Identify ports for each protocol to be removed
+	for _, endpoint := range svc.vm.Endpoints {
+		// Initialize the map only if the given protocol is encountered
+		// for the first time
+		if !encountered[endpoint.Protocol] {
+			remPorts[endpoint.Protocol] = make(map[string]bool)
+			encountered[endpoint.Protocol] = true
+		}
+		for _, port := range endpoint.Ports {
+			remPorts[endpoint.Protocol][port] = true
+		}
+	}
+
+	// Remove the the ports and protocols that are to be removed from the
+	// rules. Iterate through each item in the list of rules.
+	for indEndp := 0; indEndp < len(newRules); indEndp++ {
+		allowed := newRules[indEndp]
+		// Iterate through each port in the port list
+		for indPort := 0; indPort < len(newRules[indEndp].Ports); indPort++ {
+			port := newRules[indEndp].Ports[indPort]
+			// Process removal if the given port is to be removed
+			if remPorts[allowed.IPProtocol][port] {
+				newRules[indEndp].Ports[indPort] = ""
+				newRules[indEndp].Ports = append(
+					newRules[indEndp].Ports[:indPort],
+					newRules[indEndp].Ports[indPort+1:]...
+				)
+				// Set the index back by one because we have
+				// shifted elements from indP onwards to left
+				// by 1
+				indPort = indPort - 1
+			}
+		}
+		// At the end of removal of ports for a particular protocol if
+		// there are no ports left, then remove that protocol as well.
+		if len(newRules[indEndp].Ports) == 0 {
+			newRules[indEndp] = nil
+			newRules = append(newRules[:indEndp], newRules[indEndp+1:]...)
+			// Set the index back by one because we have shifted
+			// elements from indP onwards to left by 1
+			indEndp = indEndp - 1
+		}
+	}
+	return svc.patchFirewall(newRules)
+}
+
+// getImageList gets a list of available images in the given list of projects
```

Comment:
The libretto client will have to provide the public project. Libretto or GCP SDK does not have an API call to get the list of public projects. It can only get the list of projects that are part of the user's account.